walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,142 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (events.isEmpty() && request.getDuration() <= TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Collections.emptyList();
+    }
+    
+    List<Event> eventList = getEventsWithRequestAttendees(request, events);
+
+    if (eventList.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+ 
+    Collections.sort(eventList, Event.ORDER_BY_START);
+
+    TimeRange firstEventTimeRange = eventList.get(0).getWhen();
+    List<TimeRange> availableTimes = new ArrayList<TimeRange>();
+
+    // Add the first time slot, from the start of the day to the start of the first event,
+    // only if there's no event that starts the day.
+    if (startOfDayIsFree(eventList) && requestFits(request, firstEventTimeRange.start(), TimeRange.START_OF_DAY)) {
+      availableTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY,
+        firstEventTimeRange.start(), false));
+    }
+
+    for (int i = 0; i < eventList.size() - 1; i++) {
+      TimeRange currEventTimeRange = eventList.get(i).getWhen();
+      TimeRange nextEventTimeRange = eventList.get(i + 1).getWhen();
+
+      if (eventContainedOrOverlapped(currEventTimeRange, nextEventTimeRange)){
+        searchForNextEvent(eventList, currEventTimeRange, i, availableTimes, request);
+      } else if (requestFits(request, nextEventTimeRange.start(), currEventTimeRange.end())) {
+          availableTimes.add(TimeRange.fromStartEnd(currEventTimeRange.end(),
+            nextEventTimeRange.start(), false));  
+        }
+        
+    }
+
+    // Sort them again to be able to get the last end. Might be not that efficient.
+    Collections.sort(eventList, Event.ORDER_BY_END);
+    TimeRange lastEventTimeRange = eventList.get(eventList.size() - 1).getWhen();
+    
+    if (!availableTimes.isEmpty() && requestFits(request, TimeRange.END_OF_DAY, lastEventTimeRange.end())) {
```

Comment:
Add an inline comment explaining that part of the code. Probably you're handling the meeting being schedule at the end of the day here?